common/lib/xmodule/xmodule/modulestore/split_mongo/split.py

```diff
@@ -3034,11 +3065,18 @@ def _get_parents_from_structure(self, block_key, structure):
         Given a structure, find block_key's parent in that structure. Note returns
         the encoded format for parent
         """
-        return [
-            parent_block_key
-            for parent_block_key, value in structure['blocks'].iteritems()
-            if block_key in value.fields.get('children', [])
-        ]
+        cache_key = u'structure.{structure_id}'.format(structure_id=structure['_id'])
+        children_to_parents = cache.get(cache_key)
```

Comment:
You probably don't want this cache call here, because going out to the cache is going to be a network operation, and you don't want to do that 3K times. You'd want to build the data structure outside of this method altogether and pass it in.